store/model.go

```diff
@@ -0,0 +1,276 @@
+	ds "github.com/ipfs/go-datastore"
+	es "github.com/textileio/go-eventstore"
+)
+
+var (
+	baseKey = ds.NewKey("/model/")
+
+	ErrNotFound   = errors.New("instance not found")
+	ErrReadonlyTx = errors.New("read only transaction")
+
+	errAlreadyDiscardedCommitedTxn = errors.New("can't commit discarded/commited txn")
+	errCantCreateExistingInstance  = errors.New("can't create already existing instance")
+	errCantSaveNonExistentInstance = errors.New("can't save unkown instance")
+)
+
+type Model struct {
+	lock         sync.RWMutex
+	schema       *jsonschema.Schema
+	valueType    reflect.Type
+	datastore    ds.Datastore
+	eventcreator es.EventCreator
+	dispatcher   *es.Dispatcher
+	dsKey        ds.Key
+}
+
+func NewModel(name string, defaultInstance interface{}, datastore ds.Datastore, dispatcher *es.Dispatcher, eventcreator es.EventCreator) *Model {
+	m := &Model{
+		schema:       jsonschema.Reflect(defaultInstance),
+		datastore:    datastore,
+		valueType:    reflect.TypeOf(defaultInstance),
+		dispatcher:   dispatcher,
+		eventcreator: eventcreator,
+		dsKey:        baseKey.ChildString(name),
+	}
+
+	return m
+}
+
+func (m *Model) ReadTxn(f func(txn *Txn) error) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	txn := &Txn{model: m, readonly: true}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Model) WriteTxn(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (m *Model) FindByID(id es.EntityID, v interface{}) error {
+	return m.ReadTxn(func(txn *Txn) error {
+		return txn.FindByID(id, v)
+	})
+}
+
+func (m *Model) Add(v interface{}) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Create(v)
+	})
+}
+
+func (m *Model) Delete(id es.EntityID) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Delete(id)
+	})
+}
+
+func (m *Model) Save(v interface{}) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Save(v)
+	})
+}
+
+func (m *Model) Has(id es.EntityID) (exists bool, err error) {
+	m.ReadTxn(func(txn *Txn) error {
+		exists, err = txn.Has(id)
+		return err
+	})
+	return
+}
+
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+
+	actions []es.Action
+}
+
+type SaveOp struct {
+	Before interface{}
+	After  interface{}
+}
+
+func (t *Txn) Create(new interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
```

Comment:
Every `Create`, `Save` and `Delete` generates an `Action` in the Txn state, that will be processed on `Commit()`